File: tests/fixtures/c/sample.c
/* Sample C file for complexity analysis. */

#include <stdio.h>
#include <string.h>

#define MAX_SIZE 256

int parse_header(const char *buf, int len) {
    if (buf == NULL || len <= 0) {
        return -1;  // invalid
    }

    int total = 0;
    for (int i = 0; i < len; i++) {
        if (buf[i] == '\n') {
            continue;
        }
        if (buf[i] >= '0' && buf[i] <= '9') {
            total = total * 10 + (buf[i] - '0');
        } else {
            goto fail;
        }
    }
    return total;

fail:
    fprintf(stderr, "parse error at %d\n", total);
    return -2;
}

// compute a checksum
static unsigned crc32_small(const unsigned char *data, int n) {
    unsigned crc = 0xFFFFFFFFU;
    for (int i = 0; i < n; i++) {
        crc ^= data[i];
        for (int k = 0; k < 8; k++) {
            if (crc & 1) {
                crc = (crc >> 1) ^ 0xEDB88320U;
            } else {
                crc >>= 1;
            }
        }
    }
    return ~crc;
}

int dispatch(int op, int x) {
    switch (op) {
        case 1: return x + 1;
        case 2: return x * 2;
        case 3: return x - 1;
        default: return 0;
    }
}

File: tests/fixtures/r/sample.R
# Sample R file for complexity analysis.

MAX_SIZE <- 256

parse_header <- function(buf) {
  if (length(buf) == 0) {
    return(-1)
  }
  total <- 0
  for (b in buf) {
    if (b == 10) {
      next
    }
    if (b >= 48 && b <= 57) {
      total <- total * 10 + (b - 48)
    } else {
      stop(sprintf("parse error at %d", total))
    }
  }
  total
}

dispatch <- function(op, x) {
  if (op == 1) {
    x + 1
  } else if (op == 2) {
    x * 2
  } else if (op == 3) {
    x - 1
  } else {
    0
  }
}

crc32_small <- function(data) {
  crc <- bitwNot(0L)
  for (d in data) {
    crc <- bitwXor(crc, d)
    for (k in 1:8) {
      if (bitwAnd(crc, 1L) != 0) {
        crc <- bitwXor(bitwShiftR(crc, 1L), 0xEDB88320L)
      } else {
        crc <- bitwShiftR(crc, 1L)
      }
    }
  }
  bitwNot(crc)
}
